### ISSUE TITLE
About implementation of CameraCtrl baseline on the large VDiT

### ISSUE THREAD

**Issue by LeeKeyu** [snap-research/ac3d#28]
Hi authors, thanks a lot for your impressive work.
My question is about your implementation of CameraCtrl as a baseline to inject camera control to Video DiT, as shown in the paper.

1. Did you modify the pose encoder and linear projection layer from the original CameraCtrl paper implementation to suit for the 11B dit structure? 
2. Which dit blocks did you insert the cc_projection and camera features into? 
3. How many steps are needed to converge and the learning rate to train the cameractrl modules with the large dit.

I am currently working on implementing this method to a 14B I2V model but fail to train the camera control modules to follow the given motion signals. I suspect that the camera injection module params are too small compared to the large DiT to have any impact on the generated results. I would greatly appreciate your help!

Thanks a lot in advance!

**Comment by sherwinbahmani** [comment]
Hi,

We inserted into all DiT blocks for CameraCtrl. We increased the pose encoder dimensions to match the memory consumption of our proposed model. It also depends on your memory setup, parallelization (TP, CP, etc.) how large you can make the encoder. For CameraCtrl I was observing that the camera is not accurate, while our AC3D approach works much better while having very little amount of trainable parameters (just few millions). So the ControlNet setup of AC3D was working much better for us.

**Comment by LeeKeyu** [comment]
> Hi,
> 
> We inserted into all DiT blocks for CameraCtrl. We increased the pose encoder dimensions to match the memory consumption of our proposed model. It also depends on your memory setup, parallelization (TP, CP, etc.) how large you can make the encoder. For CameraCtrl I was observing that the camera is not accurate, while our AC3D approach works much better while having very little amount of trainable parameters (just few millions). So the ControlNet setup of AC3D was working much better for us.

@sherwinbahmani Thanks so much for your quick reply! I have a further question following your answer. May I ask did you modify the cameractrl pose encoder to have the same number of blocks as DiT, and use each intermediate feature to add into each of the DiT block? Or just use the final output of the pose encoder to inject into each DiT block?

Also, when you mean CameraCtrl is not accurate for your model, do you mean the camera is under control but not accurate or cannot control at all? Because in my experience, after modifying CameraCtrl for a DiT model, the learned model seems completely ignore the camera input, and I am so confused about the discrepancy between UNet and DiT implementation of the method. Hope you can help me with that.

Thank you a lot again for your help!

**Comment by LeeKeyu** [comment]
Could you also provide the config for training cameractrl with VDiT such as learning rate and batch size? Thank you a lot!

**Comment by sherwinbahmani** [comment]
I adapted the number of blocks to the DiT.

When I mean not accurate, it often failed for me completely. I do not have the CameraCtrl setup anymore, as it was done internally at Snap. But you should be fine using similar learning rate etc. as AC3D. Just try a few different learning rates. CameraCtrl is not directly made for DiT but for UNet, so if you can not reproduce it, you might want to try another baseline? We mainly used CameraCtrl, as there was no other camera control model out there. But now there are many camera control out and even open sourced, so why not use them?

**Comment by LeeKeyu** [comment]
> I adapted the number of blocks to the DiT.
> 
> When I mean not accurate, it often failed for me completely. I do not have the CameraCtrl setup anymore, as it was done internally at Snap. But you should be fine using similar learning rate etc. as AC3D. Just try a few different learning rates. CameraCtrl is not directly made for DiT but for UNet, so if you can not reproduce it, you might want to try another baseline? We mainly used CameraCtrl, as there was no other camera control model out there. But now there are many camera control out and even open sourced, so why not use them?

Thanks for your reply. I was using CameraCtrl as it is simple and used as baselines by many studies, so I just want to know if I adapt it wrongly so that I could not make it work with DiT. Thank you for your shared experience, and I think I will try other methods. BTW, do you think AC3D model would work well with I2V settings and what modifications might be needed? Thank you!

**Comment by sherwinbahmani** [comment]
Yes AC3D also works for I2V. The release version is only T2V but you can simply integrate it into a I2V model. Internally, the model was also implemented for I2V for the SnapVideo model.

**Comment by LeeKeyu** [comment]
@sherwinbahmani Thank you so much! Could you please give me some suggestions on modifying AC3D for HunyuanVideo structure? Since the base model is 14B vs Cogvideo 5B, would it be needed to scale up the number of layers and parameters of the controlnet?

BTW, thanks to your advice on modifying CameraCtrl, I have successfully trained a camera control model used with Hunyuan. However, I found huge degration of video quality maybe due to camera injection into all dit layers. I would greatly appreciate it if you can also share some insights on this problem i.e. how to train precise camra control without hurting the base model's video generation quality.

**Comment by sherwinbahmani** [comment]
The original SnapVideo used for AC3D was also 14B, so it should be no problem adapting it to more model parameters. I recommend injecting it into the first 25% of the layers roughly. The parameters of the ControlNet don't have to be too high. If they are too high, you get good control but the model might degrade quality as it overfits to your fine-tuning data.

What dataset are you using? Also for CameraCtrl? How long are you training?

My opinion to not degrade quality: Use some diverse video dataset annotated with camera poses or annotate diverse video data with poses using ViPE (https://research.nvidia.com/labs/toronto-ai/vipe/). 

These would be good datasets to try: https://lixsp11.github.io/sekai-project/ or https://research.nvidia.com/labs/dir/dynpose-100k/

About the model: Try to not have too many trainable parameters and not train it too long. It's important to find a good learning rate and number of training steps.

**Comment by LeeKeyu** [comment]
> The original SnapVideo used for AC3D was also 14B, so it should be no problem adapting it to more model parameters. I recommend injecting it into the first 25% of the layers roughly. The parameters of the ControlNet don't have to be too high. If they are too high, you get good control but the model might degrade quality as it overfits to your fine-tuning data.
> 
> What dataset are you using? Also for CameraCtrl? How long are you training?
> 
> My opinion to not degrade quality: Use some diverse video dataset annotated with camera poses or annotate diverse video data with poses using ViPE (https://research.nvidia.com/labs/toronto-ai/vipe/).
> 
> These would be good datasets to try: https://lixsp11.github.io/sekai-project/ or https://research.nvidia.com/labs/dir/dynpose-100k/
> 
> About the model: Try to not have too many trainable parameters and not train it too long. It's important to find a good learning rate and number of training steps.

@sherwinbahmani Thank you so much for your helpful suggestions! I am currently using realEstate10k for training, I will try the datasets you mentioned above. BTW, just to clarify, do you mean the same camera controlnet model design can work both for CogVideo (5B) and bigger base models (14B), and I only need to change the number of blocks to match the base transformer? Thank you again!

**Comment by LeeKeyu** [comment]
> The original SnapVideo used for AC3D was also 14B, so it should be no problem adapting it to more model parameters. I recommend injecting it into the first 25% of the layers roughly. The parameters of the ControlNet don't have to be too high. If they are too high, you get good control but the model might degrade quality as it overfits to your fine-tuning data.
> 
> What dataset are you using? Also for CameraCtrl? How long are you training?
> 
> My opinion to not degrade quality: Use some diverse video dataset annotated with camera poses or annotate diverse video data with poses using ViPE (https://research.nvidia.com/labs/toronto-ai/vipe/).
> 
> These would be good datasets to try: https://lixsp11.github.io/sekai-project/ or https://research.nvidia.com/labs/dir/dynpose-100k/
> 
> About the model: Try to not have too many trainable parameters and not train it too long. It's important to find a good learning rate and number of training steps.

I recently experience for CameraCtrl that injecting into the first 1/3 blocks did yield better visual quality (as also mentioned in your paper) while the camera controllability got worse, sad about this trade-off. I hope that switching to a ControlNet-based method can effectively address both issues.

**Comment by sherwinbahmani** [comment]
Yes, and the data will be important. Switching away from RealEstate10K to more diverse data will also help a lot. And make sure the ControlNet is small, like in AC3D. So not too many learnable weights.

And yes it will work for 5B and 14B, just adjust the number of blocks.

**Comment by LeeKeyu** [comment]
@sherwinbahmani Thanks so much for your helpful suggestions!! I will have a try.
